packages/automerge-repo/src/DocHandle.ts

```diff
@@ -1,181 +1,211 @@
+   * ```
+   *                        handle.state
+   * ┌───────────────┐      ┌────────────┐
+   * │new DocHandle()│  ┌──►│ LOADING    ├─┐
+   * ├─────────────┬─┘  │ ┌┤├────────────┤ │ via loadIncremental()
+   * ├─────────────┤    │ └►├────────────┤ │  or waitForSync()
+   * │find()       ├────┘ ┌─┤ REQUESTING │ │
+   * ├─────────────┤      │ ├────────────┤ │
+   * │create()     ├────┐ │ ├────────────┤ │ via receiveSyncMessage()
+   * └─────────────┘    └►└►│ READY      │►┘  or create()
+   *                        └────────────┘
+   *  ┌───────────────────┐
+   *  │value()            │ <- blocks until "ready"
+   *  ├───────────────────┤
+   *  │provisionalValue() │ <- blocks until "requesting"
+   *  └───────────────────┘
+   * ```
+   *
+   * */
+  state: HandleState = HandleState.LOADING
+
+  #log: debug.Debugger
 
   constructor(documentId: DocumentId, newDoc = false) {
     super()
     this.documentId = documentId
-    this.doc = Automerge.init({
-      patchCallback: (
-        patch: any, // Automerge.Patch,
-        before: Automerge.Doc<T>,
-        after: Automerge.Doc<T>
-      ) => this.__notifyPatchListeners(patch, before, after),
+    this.#log = debug(`ar:dochandle:${documentId}`)
+
+    this.doc = A.init({
+      patchCallback: (patch, before, after) =>
+        this.#emitPatch(patch, before, after),
     })
 
-    // new documents don't need to block on an initial value setting
-    if (newDoc) {
-      this.state = "ready"
+    // If this is a freshly created document, we can immediately mark it as ready
+    if (newDoc) this.#ready()
+  }
+
+  #ready() {
+    if (this.state !== HandleState.READY) {
+      this.state = HandleState.READY
       this.emit("ready")
     }
   }
 
-  ready() {
-    return this.state === "ready"
+  isReady() {
+    return this.state === HandleState.READY
   }
 
   loadIncremental(binary: Uint8Array) {
-    log(`[${this.documentId}]: loadIncremental`, this.doc)
-    const newDoc = Automerge.loadIncremental(this.doc, binary)
-    if (this.state === "loading") {
-      this.state = "ready"
+    if (binary.byteLength === 0) return
+    this.#log(`[${this.documentId}]: loadIncremental`, this.doc)
+    const newDoc = A.loadIncremental(this.doc, binary)
+    if (this.state === HandleState.LOADING) {
+      this.state = HandleState.READY
       this.emit("ready")
     }
-    this.__notifyChangeListeners(newDoc)
+    this.#emitChange(newDoc)
   }
 
-  unblockSync() {
-    if (this.state === "loading") {
-      this.state = "syncing"
-      this.emit("syncing")
+  /**
+   * A Repo can call this when it doesn't have the document and has advertised our interest in it.
+   * This blocks access to the document until we get it from a peer.
+   *
+   * TODO: might be good for this to timeout and go to a "not found" state if the document isn't
+   * available after a certain amount of time. but not sure what we would do with a doc in that
+   * state. We'd also need to retry etc.
+   */
+  waitForSync() {
```

Comment:
what should this be called? it doesn't actually do any waiting...
maybe `requestSync()`? 

this is kind of messy and confusing... let's move the conditional up